src/Optimizer/Transformer/RewriteAmpUrls.php

```diff
@@ -0,0 +1,356 @@
+     *
+     * @param Document        $document Document to add the meta tags to.
+     * @param string          $host     Host URL to use.
+     * @param ErrorCollection $errors   Error collection to add potential errors to.
+     */
+    private function adaptForSelfHosting(Document $document, $host, $errors)
+    {
+        // runtime-host and amp-geo-api meta tags should appear before the first script.
+        if (
+            ! $this->usesAmpCacheUrl($host)
+            &&
+            ! $this->configuration->get(RewriteAmpUrlsConfiguration::LTS)
+        ) {
+            try {
+                $url = new Url($host);
+
+                if (!empty($url->scheme) && !empty($url->host)) {
+                    $origin = "{$url->scheme}://{$url->host}";
+                    $this->addMeta($document, 'runtime-host', $origin);
+                } else {
+                    $errors->add(CannotAdaptDocumentForSelfHosting::forNonAbsoluteUrl($host));
+                }
+            } catch (Exception $exception) {
+                $errors->add(CannotAdaptDocumentForSelfHosting::fromException($exception));
+            }
+        }
+        if (
+            ! empty($this->configuration->get(RewriteAmpUrlsConfiguration::GEO_API_URL))
+            &&
+            ! $this->configuration->get(RewriteAmpUrlsConfiguration::LTS)
+        ) {
+            $this->addMeta(
+                $document,
+                'amp-geo-api',
+                $this->configuration->get(RewriteAmpUrlsConfiguration::GEO_API_URL)
+            );
+        }
+    }
+
+    /**
+     * Check whether a given URL should be preloaded.
+     *
+     * @param string $url Url to check.
+     * @return bool Whether the provided URL should be preloaded.
+     */
+    private function shouldPreload($url)
+    {
+        return substr_compare($url, 'v0.js', -5) === 0
+               ||
+               substr_compare($url, 'v0.css', -6) === 0;
+    }
+
+    /**
+     * Calculate the host string to use.
+     *
+     * @return string Host to use.
+     */
+    private function calculateHost()
+    {
+        $lts = $this->configuration->get(RewriteAmpUrlsConfiguration::LTS);
+        $rtv = $this->configuration->get(RewriteAmpUrlsConfiguration::RTV);
+
+        if ($lts && $rtv) {
+            throw InvalidConfiguration::forMutuallyExclusiveFlags(
+                RewriteAmpUrlsConfiguration::LTS,
+                RewriteAmpUrlsConfiguration::RTV
+            );
+        }
+
+        $ampUrlPrefix      = $this->configuration->get(RewriteAmpUrlsConfiguration::AMP_URL_PREFIX);
+        $ampRuntimeVersion = $this->configuration->get(RewriteAmpUrlsConfiguration::AMP_RUNTIME_VERSION);
+
+        $ampUrlPrefix = rtrim($ampUrlPrefix, '/');
+
+        if ($ampRuntimeVersion && $rtv) {
+            $ampUrlPrefix = RuntimeVersion::appendRuntimeVersion($ampUrlPrefix, $ampRuntimeVersion);
+        } elseif ($lts) {
+            $ampUrlPrefix .= '/lts';
+        }
+
+        return $ampUrlPrefix;
+    }
```

Comment:
I just realized that the `rtv` and `ampRuntimeVersion` are both empty. Instead of seeing the AMP runtime loaded with https://cdn.ampproject.org/rtv/012103261048002/v0.mjs it comes out rtv-less: https://cdn.ampproject.org/v0.mjs

However, if I hack this to force `$rtv` to be `true` and `$ampRuntimeVersion` to be `012103261048002`, then the result is invalid AMP: 

> The attribute 'src' in tag 'amphtml module engine script' is set to the invalid value 'https://cdn.ampproject.org/rtv/012103261048002/v0.mjs'

I guess versioned URLs are not yet valid AMP?

Reply:
Only `lts` is valid. RTV-specific scripts are not yet valid: https://github.com/ampproject/amphtml/blob/52407d1c72c5e2051fda427270b9503ae9900ebf/validator/validator-main.protoascii#L3109-L3256

I think this is being tracked by https://github.com/ampproject/amphtml/issues/27546, although this is about self-hosting and not allowing versioned script URLs. @sebastianbenz is that right?

Reply:
Correct, there are no plans to support RTV versions served from cdn.ampproject.org.